### PR TITLE
Features/components

### LINKED ISSUE #8: Component Outline
Here's a non-comprehensive outline of potential components that could be included in Dripsy: 

**HTML Elements** - @expo/html-elements
- H1
- H2
- H3
- H4
- H5
- H6
- H6
- A
- P
- Others?

**React Native**
- View
- ScrollView
- FlatList
- Text
- SafeAreaView
- ActivityIndicator
- ImageBackground
- Image
- KeyboardAvoidingView
- Picker
- Pressable - Needs RN 63.0
- SectionList
- StatusBar
- TouchableHighlight
- TouchableOpacity
- TouchableWithoutFeedback
- VirtualizedList
- Do we want the entire library or only specific ones?

**Theme-UI Inspired**
- Flex
- Checkbox
- Slider
- Switch
- Text
- Heading
- Badge
- Avatar
- Divider
- Progress
- Link
- Container
- Select
- Textarea
- TextInput
- SVG

**Additional**
- Segmented Control
- Modal
- Additional Components TBD

**Potential Questions**
- How much should we rely on outside dependencies? (e.i @react-native-community/checkbox, etc.) A lot might be included in Expo by default.
- Can we add a lot of components without making it too opinionated?
- Should we have heading + HTML elements, Link + A, or consolidate them somehow? 

Note that this list is non-comprehensive and any feedback on to what it should be added/removed is welcomed to improve the list.

### PR DESCRIPTION
**Pull request by cmaycumber**
# What
This PR starts adding some initial components to Dripsy. This is **not** a solidified list. Components will be both added and removed as we work to determine what is the best fit for Dripsy. Resolves #8 

# To-Do
**HTML Elements**
- [x] H1
- [x] H2
- [x] H3
- [x] H4
- [x] H5
- [x] H6
- [x] A
- [x] P
- [x] Hr
- Others?

**React Native**
- [x] View
- [x] ScrollView
- [x] FlatList
- [x] Text
- [x] SafeAreaView
- [x] ActivityIndicator
- [x] KeyboardAvoidingView
- [x] Picker
- [ ] Pressable - Needs RN 63.0
- [x] SectionList
- [x] StatusBar
- [x] TouchableHighlight
- [x] TouchableOpacity
- [x] TouchableWithoutFeedback
- [x] VirtualizedList
- Do we want the entire library or only specific ones?

**Theme-UI Inspired**
- [x] ImageBackground
- [x] Image
- [x] Flex
- [x] Row
- [x] Box
- [x] Card
- [ ] Checkbox
- [ ] Slider
- [x] Switch
- [x] Text
- [x] Heading
- [x] Badge
- [x] Avatar
- [x] Divider
- [ ] Progress
- [x] Link
- [x] Container
- [ ] Select
- [ ] Textarea
- [ ] TextInput
- [ ] SVG
- [x] Button
- [ ] Radio
- [ ] Label

### CODE REVIEW

**Comment by nandorojo** [comment]
Great - thanks. Left a few notes from a glance through.

**Comment by nandorojo** [comment]
Related for Pressable + expo: https://github.com/expo/expo/pull/9162

**Comment by nandorojo** [comment]
This is what I'm using for `Button` in my own app. Still just started it

```tsx
import { Text, Theme, useDripsyTheme, View } from 'dripsy'
import React, { ComponentProps } from 'react'
import { TouchableWithoutFeedback } from 'react-native'
import { DripsyTheme } from '../providers/Dripsy'

type Props = {
  children: React.ReactNode | string
  variant?: keyof DripsyTheme['buttons']
  sx?: ComponentProps<typeof View>['sx']
  labelSx?: ComponentProps<typeof View>['sx']
  isText?: boolean
  onPress?: () => void
  labelProps: ComponentProps<typeof Text>
  touchableProps: ComponentProps<typeof TouchableWithoutFeedback>
} & ComponentProps<typeof View>

export function Button(props: Props) {
  const {
    children,
    variant = 'primary',
    sx = {},
    labelSx = {},
    isText = true,
    onPress,
    labelProps,
    touchableProps,
    ...viewProps
  } = props

  const { theme } = useDripsyTheme()

  const { label: labelStyle = {}, ...variantStyle } = theme.buttons[
    variant
  ] ?? {
    label: {},
  }

  return (
    <TouchableWithoutFeedback {...touchableProps} onPress={onPress}>
      <View {...viewProps} sx={{ variantStyle, ...sx }}>
        {isText ? (
          <Text {...labelProps} sx={{ ...labelStyle, ...labelSx }}>{children}</Text>
        ) : (
          children
        )}
      </View>
    </TouchableWithoutFeedback>
  )
}
```

**Comment by cmaycumber** [comment]
I think that would be great to use, for now, we can always add on to it. I use something similar and also have a loading prop that renders an ActivityIndicator.

We might want to break some of the more complex components into there own file.

**Comment by nandorojo** [comment]
> We might want to break some of the more complex components into there own file.

Agreed

**Review by nandorojo — CHANGES_REQUESTED** [review]
Just realized that I commented these weeks ago but didn't hit "submit review", not sure if that means they sent or not haha

File: src/components/index.tsx [hunk 1]
```diff
@@ -54,24 +56,74 @@ export const A = createThemedComponent(HTML.A, {
   defaultVariant: 'a',
 })
 
-export const ScrollView = createThemedComponent(rScrollView)
-
-export const TextInput = createThemedComponent(rTextInput)
+export const P = createThemedComponent(HTML.P, {
+  themeKey: 'text',
+  defaultVariant: 'a',
```
[the remark this answers is not in the document]
Comment (cmaycumber) [reply src/components/index.tsx:65]: Yes, for sure. I'll switch that out.

File: src/components/index.tsx [hunk 2]
```diff
@@ -82,8 +134,40 @@ export const Container = createThemedComponent(rView, {
...
+    flexDirection: 'row',
+  },
+})
+
+export const Card = createThemedComponent(rView, {
+  themeKey: 'card',
+  defaultVariant: 'primary',
+})
+
+export const Button = createThemedComponent(rButton, {
```
[the remark this answers is not in the document]
Comment (cmaycumber) [reply src/components/index.tsx]: You're definitely right. This was more of a placeholder until we decided what we were gonna do with button.

I knew the button component would most likely have to be a little more opinionated so wanted to wait for some direction.

**Comment by cmaycumber** [comment]
I'm thinking about taking the components from either https://callstack.github.io/react-native-paper/ or https://callstack.github.io/react-native-paper/ and rebuilding them in Dripsy this weekend.

Do you think that these are too opinionated? Also what style do you think that we should be aiming for with the default Dripsy components? 

My view is taking one of these more established libraries and rebuilding them as Dripsy components will offer a lot more functionality than trying to figure it out ourselves from scratch, as well as avoid having to make development related designs related to naming conventions, what we add-in, etc.

Thoughts?

File: src/components/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (nandorojo) [reply src/components/index.tsx]: Yeah I figured all good

**Comment by nandorojo** [comment]
I'm on my phone but I'll share my thoughts I've had on this so far.

The way I see it - there are a few components, such as Buttons or badges, where we need some level of opinionated code beyond a native component. Ideally, these would be kept to a minimum and be customizable. In my opinion, you should be able to theme all of your app's core styles from your `theme` object. I resonate with the `theme-ui` goal of having a standardized theming system for the web. There are of course some times where this won't be possible, but I view it as the guiding standard.

So, to answer more specifically, I think it would be ideal for Dripsy components to stay more low-level than React Native Paper. Your styles are derived from your `theme` object, and you can build custom components from there. We could then in the docs have many example examples of theme objects, the way `theme-ui` does. I view the core of Dripsy as a responsive design system, more than a pre-made component library. 

Now, with all that said, I think using Dripsy as the building blocks for an opinionated component library is a great idea. Using the `sx` prop has been incredibly useful, and I think it should be adopted across React Native component libraries in general. What if we spun off separate packages, such as `@dripsy/material`, which use dripsy under the hood? Ideally, they would export a `theme` and a few more opinionated components that require custom code, based on the Dripsy building blocks.

(Another thing to keep in mind for this is maintenance; component libraries might have a bit of a different expectations from people who use them. Do you have to update styles over time to stay up to date with the latest? etc. I'm not too concerned about this, as long as the `theme` object handles most of the styling. That's the beauty of it.)

Hope this rambling makes sense.

**Comment by nandorojo** [comment]
Now, what I just said does make things like buttons a bit unclear. Should dripsy export a button? Probably, yeah. I think the minimal style of `theme-ui` is a nice starting point for most apps. Whether or not a custom component library would consume that same button, or create its own from scratch, I'm not exactly sure.

**Comment by cmaycumber** [comment]
> The way I see it - there are a few components, such as Buttons or badges, where we need some level of opinionated code beyond a native component. Ideally, these would be kept to a minimum and be customizable. In my opinion, you should be able to theme all of your app's core styles from your theme object. I resonate with the theme-ui goal of having a standardized theming system for the web. There are of course some times where this won't be possible, but I view it as the guiding standard.

All onboard for sticking to the theme-ui paradigm. 

> So, to answer more specifically, I think it would be ideal for Dripsy components to stay more low-level than React Native Paper. Your styles are derived from your theme object, and you can build custom components from there. We could then in the docs have many example examples of theme objects, the way theme-ui does. I view the core of Dripsy as a responsive design system, more than a pre-made component library.

Definitely the way to go. 

> Now, with all that said, I think using Dripsy as the building blocks for an opinionated component library is a great idea. Using the sx prop has been incredibly useful, and I think it should be adopted across React Native component libraries in general. What if we spun off separate packages, such as @dripsy/material, which use dripsy under the hood? Ideally, they would export a theme and a few more opinionated components that require custom code, based on the Dripsy building blocks.

I agree with just about everything you said. The idea of mini style-specific libraries is great like @dripsy/material is great. I think that's much better than what I proposed. That will give us tremendous flexibility to experiment and also have a wide range of different themes/styles.

Overall, I think this is great. I'll try to replicate some of these concepts going forward with this branch and hopefully get something out by the end of the weekend.

**Comment by nandorojo** [comment]
Tailwind UI might be my favorite inspiration I've come across for a UI library. Very clean and customizable.

https://tailwindui.com/components#product-application-ui

**Comment by cmaycumber** [comment]
https://blog.expo.io/expo-sdk-39-is-now-available-4c10aa825e3f This should be relevant for creating a Button with the pressable component.

**Comment by nandorojo** [comment]
Oh thanks for sharing I hadn't seen that yet. Great release overall, and good to see the Pressable. Funny to see my feature request for screenshot detection get linked to.

They haven't upgraded to RNW 14 yet which will include the new Pressable with hover support, but I've mentioned this to their team and assume it'll get upgraded soon.

For now, I'll probably still use `react-native-web-hooks` for hovering styles

**Comment by nandorojo** [comment]
Another really nice set of components from Vercel: https://react.geist-ui.dev/en-us/components/button

**Comment by cmaycumber** [comment]
I took a quick look at geist and really the library as a whole. Specifically, I'm a fan of how it seems very comprehensive at the same time without being too opinionated. I personally component libraries with a lot of components. Couple something like that with the themeability that Dripsy offers I think something like that would be awesome.

**Comment by cmaycumber** [comment]
Closing this temporarily as we transition to monorepo in #48

**Comment by nandorojo** [comment]
I've been slowly making a geist-like dripsy package inside of my own app. I've been going component-by-component as I need them, and will move towards putting it in its own package once it feels robust enough.

Here's an example of a component I made following the geist style patterns: https://snack.expo.io/@nandorojo/dripsy-browser-mockup

**Pull request closed without merging**